feat #4: 사용자 설정 업데이트 API 구현

piopoi opened the pull request:
## 📃 설명

- resolves #4 
- 사용자 정보 업데이트 기능 추가
  - 업데이트 가능한 정보: 위도, 경도, 추천사용여부
  - 로그인 한 사용자 본인의 정보만 수정 가능함.


## 🔨 작업 내용

- 사용자 정보 업데이트 API 추가
- **Exception 처리 추가**
  - 과제모음 > 지리기반 맛집 추천 웹 서비스 > 트러블 슈팅 > [Exception 처리](https://www.notion.so/Exception-78135539fbde466996ff1ed6c951060b?pvs=4)

## 💬 기타 사항

-

piopoi committed:
feat #4: 사용자 설정 업데이트 API 구현

piopoi committed:
refactor #4: 위도, 경도 자료형 변경(BigDecimal -> String)

piopoi committed:
fix #4: 사용자 생성자 파라미터 수정

piopoi committed:
Merge remote-tracking branch 'origin/develop' into feat/#4-api-user-update-location

piopoi committed:
feat #4: Global Exception 처리 추가

piopoi commented:
공통 Exception 처리 추가하였으니 참고하시기 바랍니다.

노션페이지에도 간단하게 내용 추가해두었습니다.
과제모음 > 지리기반 맛집 추천 웹 서비스 > 트러블 슈팅 > [Exception 처리](https://www.notion.so/Exception-78135539fbde466996ff1ed6c951060b?pvs=4)

SteadyFlow736 commented on src/test/java/com/wanted/teamr/tastyfinder/api/member/controller/MemberControllerTest.java:
```diff
@@ -105,6 +135,84 @@ void createMember_emptyPassword() throws Exception {
+    }
+
+    @Test
+    @WithUserDetails(value = testEmail, setupBefore = TestExecutionEvent.TEST_EXECUTION)
```
오... setupBefore 부분이 있는 줄 몰랐습니다. 타이밍을 세밀하게 맞출 수 있는 것 같네요. TEST_EXECUTION이면  혹시 BeforeEach가 붙은 setup 실행 이후로  UserDetails를 가져오게 되나요?

piopoi replied:
@Before 호출 -> @WithUserDetails 호출 -> 테스트 메서드 호출 순으로 진행되는 것 같습니다.

[@WithUserDetails 공식 문서](https://docs.spring.io/spring-security/reference/servlet/test/method.html#test-method-withuserdetails)의 내용을 보시면, @WithUserDetails 항목 마지막에 관련된 문단이 있습니다.
번역하면 아래와 같습니다. (with DeepL)
> 기본적으로 SecurityContext는 TestExecutionListener.beforeTestMethod 이벤트 중에 설정됩니다. 이는 JUnit의 @Before 이전에 발생하는 것과 동일합니다. 이를 변경하여 테스트 메서드가 호출되기 전, 즉 JUnit의 @Before 이후인 TestExecutionListener.beforeTestExecution 이벤트 중에 발생하도록 할 수 있습니다:
`@WithUserDetails(setupBefore = TestExecutionEvent.TEST_EXECUTION)`

mojh7 replied:
이번에 `@WithMockUser` 도 처음봐서 신기했는데 `WithUserDetails`도 있다고 얘기하셔서 신기하고 옵션도 다양하게 있어서 더 신기하네요

SteadyFlow736 replied:
아하 그렇다면 setup에서 Member가 저장되고 이후 @WithUser 호출되어 저장된 멤버로 SecurityContext가 셋업되는 것이겠군요. Security 관련된 테스트 방법이 이렇게 다양한 줄 몰랐습니다.

sohyuneeee replied:
`@WithUserDetails` 라는게 있는지 몰랐는데 덕분에 알아갑니다!!! 이 방법으로 테스트코드 구현해볼게요!!!

JoonheeJeong replied:
ㅋㅋㅋ 다들 감탄하고가는 포인트네요~ 저도 잘 봤습니다! 나중에 직접 테스트 만들어보면서 경험해서 제 걸로 흡수해야겠네요!! ㅎㅎ

SteadyFlow736 commented on src/main/java/com/wanted/teamr/tastyfinder/api/member/validation/LongitudeValidator.java:
ConstraintValidator를 이용한 커스텀 validator가 참 유용한 것 같습니다. 👍

SteadyFlow736 commented on src/main/java/com/wanted/teamr/tastyfinder/api/exception/GlobalExceptionHandler.java:
```diff
@@ -15,16 +21,60 @@
+        try {
+            return valueOf(firstErrorMessage);
+        } catch (IllegalArgumentException | NullPointerException ex) {
+            return COMMON_INVALID_PARAMETER;
```
오 이런 방법이! 👍
적어도 이 메시지를 받은 사용자 또는 프론트엔드 엔지니어가 우리에게 메시지에 대해 문의한다면, constraint error message를 잘못 적은 실수를 알 수 있겠네요. 물론 그 전에 자동 테스트가 1차 방어선이 되어야 할 것 같습니다. 😀

piopoi replied:
저 부분에 대해 고민을 많이 했는데, exception을 발생시킬까 하다가 기본값을 추가해주는 방향으로 갔습니다.
@ControllerAdvice 내에서 Exception 발생시면 @ControllerAdvice로 핸들링이 안되서 response에 message 세팅하기가 어렵더라구요. 😄

JoonheeJeong replied:
이 방법이 지금까지 방법중에 가장 탁월하네요!

JoonheeJeong replied:
하드코딩이 있긴 하지만 이 방법이 지금까지 본 것 중에 가장 탁월한 것 같네요!

SteadyFlow736 reviewed (approved):
석원님 코드 잘 봤습니다. MethodArgumentNotValidException에 대한 대응과 커스텀 validator가 인상적이었습니다. 👍

mojh7 commented on src/main/java/com/wanted/teamr/tastyfinder/api/member/dto/MemberUpdateRequest.java:
```diff
@@ -0,0 +1,22 @@
+@Builder
+public class MemberUpdateRequest {
+
+    @DecimalMax("38.61") //대한민국 최북단 북위 38.61
```
맛집 목록 조회할 때 위도 경도가 필수값이라 null 체크만 생각 하고 범위 지정할 생각은 못했는데
회의 때 얘기하신 대로 이렇게 하는게 좋은 것 같습니다 👍 👍

piopoi replied:
엇 저 부분은 처음에 커밋했던 내용이네요.
지금은 validator 사용한 annotation을 별도로 생성하여 체크하고 있습니다.

mojh7 replied:
넵 다 보고나니 validator로 옮겨진 것 확인 했었는데
리뷰 달 때는 처음 커밋 부터 읽었어서 저기에다가 리뷰 달았나 봅니다

piopoi committed:
feat #4: 인증/인가 exception handler 추가

- 인증 exception handler: AuthenticationEntryPointImpl
- 인가 exception handler: AccessDeniedHandlerImpl

sohyuneeee commented on src/main/java/com/wanted/teamr/tastyfinder/api/member/dto/MemberCreateRequest.java:
```diff
@@ -9,10 +9,11 @@
-    @NotBlank
-    @Email(regexp = "^[\\w!#$%&amp;'*+/=?`{|}~^-]+(?:\\.[\\w!#$%&amp;'*+/=?`{|}~^-]+)*@(?:[a-zA-Z0-9-]+\\.)+[a-zA-Z]{2,6}$")
+    @NotBlank(message = "MEMBER_EMAIL_EMPTY")
+    @Email(message = "MEMBER_EMAIL_INVALID",
+            regexp = "^[\\w!#$%&amp;'*+/=?`{|}~^-]+(?:\\.[\\w!#$%&amp;'*+/=?`{|}~^-]+)*@(?:[a-zA-Z0-9-]+\\.)+[a-zA-Z]{2,6}$")
     private String email;
 
-    @NotBlank
+    @NotBlank(message = "MEMBER_PASSWORD_EMPTY")
```
validation annotation 의 message 를 설정해줄 수 있는지 몰랐는데 덕분에 알고 갑니다!!!

sohyuneeee reviewed (approved):
석원님ㅎㅎ 덕분에 이번에도 많은 걸 알아갑니다! 감사합니다!! 고생많으셨습니다!

mojh7 commented on src/main/java/com/wanted/teamr/tastyfinder/api/exception/GlobalExceptionHandler.java:
```diff
@@ -60,23 +60,23 @@ private String getFirstErrorMessage(MethodArgumentNotValidException e) {
                 .body(customErrorResponse);
     }
+
+    private void logError(Exception e) {
```
이 작업 따로 함수로 묶을 생각은 못 했는데 이렇게 하는게 좋아보입니다

piopoi replied:
이전까지는 메서드명을 문자열로 직접 입력해서 저렇게 따로 만들지는 않았는데,
하드코딩 줄이려고 `e.getClass().getSimpleName()` 이렇게 쓰다보니 같은 코드가 반복되서 따로 메서드로 생성해주었습니다.

JoonheeJeong replied:
메소드 이렇게 나눠서 리팩토링 해주신 거 정말 좋네요!

JoonheeJeong replied:
반복되는 로직들 별도의 메소드로 잘게 리팩토링 해주신 것 너무 좋네요!! 가독성이 훨씬 좋아지는 것 같습니다! 👍🏼
(기존 리뷰에 붙인다고 이 부분에 한정되었는데, 예외처리기 클래스 자체에서 전반적으로 리팩토링을 잘해주신 것 같습니다!!)

mojh7 reviewed (approved):
회의 때도 이번 pr 관련 얘기 들은 것도 있고 notion에 써주신 것도 있어서
뭔가 리뷰 따로 달게 없네요

처음에 사실 custom vaildator 어떻게 하나 모를 때 생각으로는 아래처럼 됐으면 했는데
`@NotBlank(우리가 만든 예외 정보 넘기기)`

회의 때 enum 다음에 함수 아무것도 자동 완성 안되는 것 보고 아차했습니다
결국 어디선가는 저희가 직접 넣어야 되는군요

그래도 직접 입력하면 생길만한 위험 요소를 GlobalExceptionHandler에서 한 번 막기도
하고 상세한 예외 필드 정보는 못 줘도 파라미터가 잘 못 입력했다는 정보를 프론트에게 넘길 수
있어서 좋은 것 같습니다

실제로 이렇게 서비스한다 가정해도 `COMMON_INVALID_PARAMETER` 예외 나오는지
모니터링만 잘 해놓고 발견될 때 수정해도 크게 문제가 없는 것 같아 좋은 것 같습니다
고생하셨습니다 👍 👍

JoonheeJeong commented on src/main/java/com/wanted/teamr/tastyfinder/api/exception/AccessDeniedHandlerImpl.java:
```diff
@@ -0,0 +1,36 @@
+                       AccessDeniedException accessDeniedException) throws IOException, ServletException {
+        log.error(accessDeniedException.getClass().getSimpleName(), accessDeniedException.getMessage());
+
+        CustomErrorResponse customErrorResponse = CustomErrorResponse.of(AUTH_AUTHORIZATION_FAILED);
+        String responseBody = new ObjectMapper().writeValueAsString(customErrorResponse);
```
요거 소현님 코드도 마찬가지였는데, ObjectMapper가 핸들링마다 새롭게 생성되고 있어서 자동주입 받을 수 있으면 받는 게 좋을 것 같더라구요! 한 번 확인 부탁드립니다! (`AuthenticationEntryPointImpl` 도 마찬가지입니다!)

piopoi replied:
Security 쪽에서 말씀하신 부분들 수정하였습니다.
좋은 코멘트 감사합니다. 😄

JoonheeJeong commented on src/main/java/com/wanted/teamr/tastyfinder/api/member/controller/MemberController.java:
```diff
@@ -21,4 +25,12 @@ public ResponseEntity<Void> createMember(@RequestBody @Valid MemberCreateRequest
+    @PatchMapping("/api/members")
+    public ResponseEntity<Void> updateMember(@AuthenticationPrincipal MemberAdapter memberAdapter,
+                                             @RequestBody @Valid MemberUpdateRequest memberUpdateRequest) {
+        Long memberId = memberAdapter.getMember().getId();
+        memberService.updateMember(memberId, memberUpdateRequest);
+        return ResponseEntity.ok().build();
+    }
```
서비스 전용 dto 만들지 않고 일단 Id랑 컨트롤러 dto 그대로 넘겨주는 거 좋은 것 같습니다~! 프로젝트 커짐에 따라 점진적으로 dto를 더 세분화하는 게 좋다는 저희 저번에 논의가 기억이 나네요 ㅋㅋ!

piopoi replied:
MSA로 바뀌게 되면 자연스럽게 레이어 별 DTO를 만들게 될 것 같습니다.
현재와 같은 구조는 Controller layer와 Service layer의 결합도가 높아져서, 다른 모듈에서 서비스의 기능(메서드)을 호출하기 어려워지는 단점이 있더라구요.

JoonheeJeong commented on src/main/java/com/wanted/teamr/tastyfinder/api/exception/GlobalExceptionHandler.java:
```diff
@@ -15,18 +21,62 @@
+        BindingResult bindingResult = e.getBindingResult();
+        return bindingResult
+                .getAllErrors()
+                .get(FIRST_ERROR_INDEX) //첫 번째 에러만 반환
+                .getDefaultMessage();
```
사소할 수 있는데 상수로 빼신 것 굿입니다 ㅎㅎ

JoonheeJeong commented on src/main/java/com/wanted/teamr/tastyfinder/api/exception/CustomErrorResponse.java:
```diff
@@ -1,15 +1,24 @@
+        this.message = message;
+    }
+
+    public static CustomErrorResponse of(ErrorCode errorCode) {
+        return CustomErrorResponse.builder()
+                .code(errorCode.name())
+                .message(errorCode.getMessage())
+                .build();
     }
```
사소할 수 있지만, 저번에 보여주셨던 [링크](https://tecoble.techcourse.co.kr/post/2020-05-26-static-factory-method/)에서 인자가 하나인 경우는 `from`을 쓰는 게 일반적인것 같더라구요. 사실 명명규칙이야 팀에서 정하기 나름이니까 상관 없긴 한데, 생각나서 말씀드려봤습니다!

piopoi replied:
저도 from으로 쓰고 싶었는데, 저희 컨벤션이 of로 통일하자고 하여 of로 네이밍했습니다. 😭 

추가적으로 전 일반적으로 사용하는 네이밍 규칙은 웬만하면 지켜주는게 좋다고 생각해요.
팀에 새로운 개발자가 들어오거나 다른 팀의 개발자가 코드를 보거나 할 때, 
익숙한 네이밍으로 더 빠르게 볼 수 있다는 게 의외로 많은 도움이 되더라구요.

JoonheeJeong commented on src/main/java/com/wanted/teamr/tastyfinder/api/member/dto/MemberUpdateRequest.java:
```diff
@@ -0,0 +1,19 @@
+@Builder
+public class MemberUpdateRequest {
+
+    @Latitude(message = "MEMBER_LATITUDE_INVALID")
+    private String latitude; //위도
+
+    @Longitude(message = "MEMBER_LONGITUDE_INVALID")
+    private String longitude; //경도
```
커스텀 애너테이션에 메시지 default값을 지금 입력해주신 값으로 지정하고 입력생략하는 것도 좋아보입니다~! 다시봐도 위/경도 커스텀 애너테이션 너무 잘 만드셨네요..👍🏼

JoonheeJeong commented on src/main/java/com/wanted/teamr/tastyfinder/api/member/service/MemberService.java:
```diff
@@ -20,4 +21,15 @@ public Long createMember(MemberCreateRequest memberCreateRequest) {
+        member.update(memberUpdateRequest);
+    }
+
+    private Member findMemberById(Long memberId) {
+        return memberRepository.findById(memberId)
+                .orElseThrow(() -> new IllegalArgumentException("존재하지 않는 사용자입니다."));
+    }
```
요거 커스텀 예외로 빼주시면 좋겠네요!

piopoi replied:
앗 이 부분은 #5 에서 커스텀 예외로 수정된 부분입니다.
다음 PR에서 수정될거에요 😄

JoonheeJeong reviewed (commented):
와 석원님 코드 보면 정말 배우는 게 많은 것 같아요! 리뷰 즐겁게 잘했습니다!

대부분은 참고사항인데 ObjectMapper 주입 등 꼭 확인해주셨으면 하는 게 있어서 approve 말고 코멘트로 남깁니다!

piopoi committed:
refactor #4: SecurityConfig에서 new를 이용한 객체 생성 대신 bean으로 관리되는 객체 사용하도록 수정
